--- a/compat/Makefile.am
+++ b/compat/Makefile.am
@@ -1,7 +1,7 @@
 compatdir = $(xkb_base)/compat
 
 compat_DATA = \
-accessx basic caps complete \
+accessx basic caps complete chromebook \
 iso9995 \
 japan ledcaps \
 lednum ledscroll level5 \
--- /dev/null
+++ b/compat/chromebook
@@ -0,0 +1,7 @@
+// Overlay1_Enable is a latch key for overlay1
+
+default partial xkb_compatibility "overlay"  {
+    interpret Overlay1_Enable+AnyOfOrNone(all) {
+        action= SetControls(controls=Overlay1);
+    };
+};
--- a/keycodes/Makefile.am
+++ b/keycodes/Makefile.am
@@ -6,6 +6,7 @@
 aliases \
 amiga \
 ataritt \
+chromebook_m \
 empty \
 evdev \
 fujitsu \
--- /dev/null
+++ b/keycodes/chromebook_falco
@@ -0,0 +1,21 @@
+default xkb_keycodes "fkeys" {
+	<FK01> = 166; //Back
+	<FK02> = 167; //Forward
+	<FK03> = 181; //Refresh
+	<FK05> = 235; //Video mode
+	<FK06> = 232; //Brightness down
+	<FK07> = 233; //Brightness up
+	<FK08> = 121; //Mute
+	<FK09> = 122; //Volume -
+	<FK10> = 123; //Volume +
+
+        <I166> = 67;   // #define KEY_BACK
+        <I167> = 68;   // #define KEY_FORWARD
+        <I181> = 69;   // #define KEY_REFRESH
+        <I235> = 71;   // #define KEY_SWITCHVIDEOMODE
+        <I232> = 72;   // #define KEY_BRIGHTNESSDOWN
+        <I233> = 73;   // #define KEY_BRIGHTNESSUP
+        <MUTE> = 74;
+        <VOL-> = 75;
+        <VOL+> = 76;
+};
--- /dev/null
+++ b/keycodes/chromebook_m
@@ -0,0 +1,28 @@
+default xkb_keycodes "media" {
+ # These lines remap media keys to F keys, why?
+ #	<FK01> = 166; //Back
+ #	<FK02> = 167; //Forward
+ #	<FK03> = 181; //Refresh
+ #	<FK04> = 160; //Video next
+ #	<FK05> = 235; //Video mode
+ #	<FK06> = 232; //Brightness down
+ #	<FK07> = 233; //Brightness up
+ #	<FK08> = 121; //Mute
+ #	<FK09> = 122; //Volume -
+ #	<FK10> = 123; //Volume +
+
+        <I166> = 67;   // #define KEY_BACK
+        <I167> = 68;   // #define KEY_FORWARD
+        <I181> = 69;   // #define KEY_REFRESH
+        <I160> = 70;   // #define KEY_COFFEE
+        <I235> = 71;   // #define KEY_SWITCHVIDEOMODE
+        <I232> = 72;   // #define KEY_BRIGHTNESSDOWN
+        <I233> = 73;   // #define KEY_BRIGHTNESSUP
+        <MUTE> = 74;
+        <VOL-> = 75;
+        <VOL+> = 76;
+        # Volume rocker for convertible models
+        <VLR-> = 122;
+        <VLR+> = 123;
+
+};
--- /dev/null
+++ b/keycodes/chromebook_m_falco
@@ -0,0 +1,4 @@
+default xkb_keycodes "media" {
+	<FK04> = 160; //Video next
+        <I160> = 70;  // #define KEY_COFFEE
+};
--- a/rules/base.ml1_c.part
+++ b/rules/base.ml1_c.part
@@ -1 +1,10 @@
-  *		*		=	complete
+  chromebook              *       =       complete+chromebook
+  chromebook_ralt         *       =       complete+chromebook
+  chromebook_m            *       =       complete+chromebook
+  chromebook_m_ralt       *       =       complete+chromebook
+  chromebook_falco        *       =       complete+chromebook
+  chromebook_falco_ralt   *       =       complete+chromebook
+  chromebook_m_falco      *       =       complete+chromebook
+  chromebook_m_falco_ralt *       =       complete+chromebook
+  chromebook_no_m         *       =       complete
+  *		          *	  =    	  complete
--- a/rules/base.ml_c.part
+++ b/rules/base.ml_c.part
@@ -1,5 +1,14 @@
-  pc98		nec_vndr/jp	=	pc98(basic)
-  *		jp		=	complete+japan
-  olpc          *               =       olpc
-  olpcm         *               =       olpc
-  *		*		=	complete
+  chromebook              *               =       complete+chromebook
+  chromebook_ralt         *               =       complete+chromebook
+  chromebook_m            *               =       complete+chromebook
+  chromebook_m_ralt       *               =       complete+chromebook
+  chromebook_falco        *               =       complete+chromebook
+  chromebook_falco_ralt   *               =       complete+chromebook
+  chromebook_m_falco      *               =       complete+chromebook
+  chromebook_m_falco_ralt *               =       complete+chromebook
+  chromebook_no_m         *               =       complete
+  pc98		          nec_vndr/jp     =       pc98(basic)
+  *		          jp	          =       complete+japan
+  olpc                    *               =       olpc
+  olpcm                   *               =       olpc
+  *		          *	          =       complete
--- a/rules/base.xml.in
+++ b/rules/base.xml.in
@@ -4,6 +4,69 @@
   <modelList>
     <model>
       <configItem>
+        <name>chromebook</name>
+        <description>Chromebook (most models) | Search overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_ralt</name>
+        <description>Chromebook (most models) | Right alt overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m</name>
+        <description>Chromebook (most models) | Search overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_ralt</name>
+        <description>Chromebook (most models) | Right alt overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_falco</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Search overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_falco_ralt</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Right alt overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_falco</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Search overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_falco_ralt</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Right alt overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_no_m</name>
+        <description>Chromebook (most models) | No overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
         <name>pc101</name>
         <_description>Generic 101-key PC</_description>
         <vendor>Generic</vendor>
--- a/rules/evdev.m_k.part
+++ b/rules/evdev.m_k.part
@@ -2,4 +2,10 @@
   applealu_jis	=	evdev+macintosh(jisevdev)
   olpc          =       evdev+olpc(olpc)
   olpcm         =       evdev+olpc(olpcm)
+  chromebook_m  =       evdev+chromebook_m(media)
+  chromebook_m_ralt  =  evdev+chromebook_m(media)
+  chromebook_falco_m  = evdev+chromebook_m_falco(media)
+  chromebook_falco  = evdev+chromebook_falco(fkeys)
+  chromebook_falco_ralt  = evdev+chromebook_falco(fkeys)
+  chromebook_no_m  =    evdev+chromebook_m(media)
   *		=	evdev
--- a/rules/evdev.m_s.part
+++ b/rules/evdev.m_s.part
@@ -1,4 +1,12 @@
   $evdevkbds    =   +inet(evdev)+inet(%m)
-  chromebook    =   +inet(evdev)+inet(chromebook)
+  chromebook               =   +inet(evdev)+chromebook(overlay)
+  chromebook_ralt          =   +inet(evdev)+chromebook_ralt(overlay)
+  chromebook_m             =   +inet(evdev)+chromebook_m(overlay)
+  chromebook_m_ralt        =   +inet(evdev)+chromebook_m_ralt(overlay)
+  chromebook_falco         =   +inet(evdev)+chromebook(overlay)
+  chromebook_falco_ralt    =   +inet(evdev)+chromebook_ralt(overlay)
+  chromebook_m_falco       =   +inet(evdev)+chromebook_m(overlay)
+  chromebook_m_falco_ralt  =   +inet(evdev)+chromebook_m_ralt(overlay)
+  chromebook_no_m          =   +inet(evdev)
   applealu_jis  =   +inet(evdev)+macintosh_vndr/jp(alujiskeys)
   *             =   +inet(evdev)
--- a/rules/evdev.xml.in
+++ b/rules/evdev.xml.in
@@ -1309,6 +1309,69 @@
        <vendor>Megawin Technology</vendor>
      </configItem>
     </model>
+    <model>
+      <configItem>
+        <name>chromebook</name>
+        <description>Chromebook (most models) | Search overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_ralt</name>
+        <description>Chromebook (most models) | Right alt overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m</name>
+        <description>Chromebook (most models) | Search overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_ralt</name>
+        <description>Chromebook (most models) | Right alt overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_falco</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Search overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_falco_ralt</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Right alt overlay</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_falco</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Search overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_m_falco_ralt</name>
+        <description>Chromebook Falco/Pixel/Pixel2 | Right alt overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
+        <name>chromebook_no_m</name>
+        <description>Chromebook (most models) | No overlay | F keys mapped to media keys</description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
   </modelList>
   <layoutList>
     <layout>
--- a/symbols/Makefile.am
+++ b/symbols/Makefile.am
@@ -34,6 +34,7 @@
 ua us uz vn \
 za \
 altwin capslock compose ctrl empty eurosign rupeesign group inet \
-keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo
+keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo chromebook \
+chromebook_m chromebook_m_ralt chromebook_ralt
 
 EXTRA_DIST = $(symbols_DATA)
--- /dev/null
+++ b/symbols/chromebook
@@ -0,0 +1,49 @@
+// This mapping assumes that inet(evdev) will also be sourced
+partial
+xkb_symbols "overlay" {
+    key <LWIN> { [ Overlay1_Enable ], overlay1=<LWIN> };
+    key <I225> { [ Overlay1_Enable ], overlay1=<I225> };
+
+    key <AB09> { overlay1=<INS> };
+
+    key <LEFT> { overlay1=<HOME> };
+    key <RGHT> { overlay1=<END> };
+    key <UP>   { overlay1=<PGUP> };
+    key <DOWN> { overlay1=<PGDN> };
+
+    key <FK01> { overlay1=<I166> };
+    key <FK02> { overlay1=<I167> };
+    key <FK03> { overlay1=<I181> };
+    key <FK04> { overlay1=<I160> };
+    key <FK05> { overlay1=<I235> };
+    key <FK06> { overlay1=<I232> }; // XF86MonBrightnessDown
+    key <FK07> { overlay1=<I233> }; // XF86MonBrightnessUp
+    key <FK08> { overlay1=<MUTE> };
+    key <FK09> { overlay1=<VOL-> };
+    key <FK10> { overlay1=<VOL+> };
+
+    key <AE01> { overlay1=<FK01> };
+    key <AE02> { overlay1=<FK02> };
+    key <AE03> { overlay1=<FK03> };
+    key <AE04> { overlay1=<FK04> };
+    key <AE05> { overlay1=<FK05> };
+    key <AE06> { overlay1=<FK06> };
+    key <AE07> { overlay1=<FK07> };
+    key <AE08> { overlay1=<FK08> };
+    key <AE09> { overlay1=<FK09> };
+    key <AE10> { overlay1=<I236> }; // KEY_KBDILLUMTOGGLE
+    key <AE11> { overlay1=<I237> }; // KEY_KBDILLUMDOWN
+    key <AE12> { overlay1=<I238> }; // KEY_KBDILLUMUP
+    key <BKSP> { overlay1=<DELE> };
+
+    key <LALT> { overlay1=<CAPS> };
+    key <RALT> { overlay1=<CAPS> };
+
+    // For some strange reason, some Super_R events are triggered when
+    // the Search key is released (i.e. with overlay on).
+    // This maps RWIN to a dummy key (<I253>), to make sure we catch it.
+    key <RWIN> { [ NoSymbol ], overlay1=<I253> };
+
+    // Map dummy key to no symbol
+    key <I253> { [ NoSymbol ] };
+};
--- /dev/null
+++ b/symbols/chromebook_m
@@ -0,0 +1,53 @@
+// This mapping assumes that inet(evdev) will also be sourced
+partial
+xkb_symbols "overlay" {
+    key <LWIN> { [ Overlay1_Enable ], overlay1=<LWIN> };
+    key <I225> { [ Overlay1_Enable ], overlay1=<I225> };
+
+    key <AB09> { overlay1=<INS> };
+
+    key <LEFT> { overlay1=<HOME> };
+    key <RGHT> { overlay1=<END> };
+    key <UP>   { overlay1=<PGUP> };
+    key <DOWN> { overlay1=<PGDN> };
+
+    key <I166> { overlay1=<FK01> };
+    key <I167> { overlay1=<FK02> };
+    key <I181> { overlay1=<FK03> };
+    key <I160> { overlay1=<FK04> };
+    key <I235> { overlay1=<FK05> };
+    key <I232> { overlay1=<FK06> };
+    key <I233> { overlay1=<FK07> };
+    key <MUTE> { overlay1=<FK08> };
+    key <VOL-> { overlay1=<FK09> };
+    key <VOL+> { overlay1=<FK10> };
+
+    # Volume rocker for convertibles
+    key <VLR-> { [ XF86AudioLowerVolume ] };
+    key <VLR+> { [ XF86AudioRaiseVolume ] };
+
+    key <AE01> { overlay1=<FK01> };
+    key <AE02> { overlay1=<FK02> };
+    key <AE03> { overlay1=<FK03> };
+    key <AE04> { overlay1=<FK04> };
+    key <AE05> { overlay1=<FK05> };
+    key <AE06> { overlay1=<FK06> };
+    key <AE07> { overlay1=<FK07> };
+    key <AE08> { overlay1=<FK08> };
+    key <AE09> { overlay1=<FK09> };
+    key <AE10> { overlay1=<I236> }; // KEY_KBDILLUMTOGGLE
+    key <AE11> { overlay1=<I237> }; // KEY_KBDILLUMDOWN
+    key <AE12> { overlay1=<I238> }; // KEY_KBDILLUMUP
+    key <BKSP> { overlay1=<DELE> };
+
+    key <LALT> { overlay1=<CAPS> };
+    key <RALT> { overlay1=<CAPS> };
+
+    // For some strange reason, some Super_R events are triggered when
+    // the Search key is released (i.e. with overlay on).
+    // This maps RWIN to a dummy key (<I253>), to make sure we catch it.
+    key <RWIN> { [ NoSymbol ], overlay1=<I253> };
+
+    // Map dummy key to no symbol
+    key <I253> { [ NoSymbol ] };
+};
--- /dev/null
+++ b/symbols/chromebook_m_ralt
@@ -0,0 +1,43 @@
+// This mapping assumes that inet(evdev) will also be sourced
+partial
+xkb_symbols "overlay" {
+    key <RALT> { [ Overlay1_Enable ], overlay1=<RALT> };
+
+    key <AB09> { overlay1=<INS> };
+
+    key <LEFT> { overlay1=<HOME> };
+    key <RGHT> { overlay1=<END> };
+    key <UP>   { overlay1=<PGUP> };
+    key <DOWN> { overlay1=<PGDN> };
+
+    key <I166> { overlay1=<FK01> };
+    key <I167> { overlay1=<FK02> };
+    key <I181> { overlay1=<FK03> };
+    key <I160> { overlay1=<FK04> };
+    key <I235> { overlay1=<FK05> };
+    key <I232> { overlay1=<FK06> };
+    key <I233> { overlay1=<FK07> };
+    key <MUTE> { overlay1=<FK08> };
+    key <VOL-> { overlay1=<FK09> };
+    key <VOL+> { overlay1=<FK10> };
+
+    # Volume rocker for convertibles
+    key <VLR-> { [ XF86AudioLowerVolume ] };
+    key <VLR+> { [ XF86AudioRaiseVolume ] };
+
+    key <AE01> { overlay1=<FK01> };
+    key <AE02> { overlay1=<FK02> };
+    key <AE03> { overlay1=<FK03> };
+    key <AE04> { overlay1=<FK04> };
+    key <AE05> { overlay1=<FK05> };
+    key <AE06> { overlay1=<FK06> };
+    key <AE07> { overlay1=<FK07> };
+    key <AE08> { overlay1=<FK08> };
+    key <AE09> { overlay1=<FK09> };
+    key <AE10> { overlay1=<I236> }; // KEY_KBDILLUMTOGGLE
+    key <AE11> { overlay1=<I237> }; // KEY_KBDILLUMDOWN
+    key <AE12> { overlay1=<I238> }; // KEY_KBDILLUMUP
+    key <BKSP> { overlay1=<DELE> };
+
+    key <RTSH> { overlay1=<CAPS> };
+};
--- /dev/null
+++ b/symbols/chromebook_ralt
@@ -0,0 +1,39 @@
+// This mapping assumes that inet(evdev) will also be sourced
+partial
+xkb_symbols "overlay" {
+    key <RALT> { [ Overlay1_Enable ], overlay1=<RALT> };
+
+    key <AB09> { overlay1=<INS> };
+
+    key <LEFT> { overlay1=<HOME> };
+    key <RGHT> { overlay1=<END> };
+    key <UP>   { overlay1=<PGUP> };
+    key <DOWN> { overlay1=<PGDN> };
+
+    key <FK01> { overlay1=<I166> };
+    key <FK02> { overlay1=<I167> };
+    key <FK03> { overlay1=<I181> };
+    key <FK04> { overlay1=<I160> };
+    key <FK05> { overlay1=<I235> };
+    key <FK06> { overlay1=<I232> }; // XF86MonBrightnessDown
+    key <FK07> { overlay1=<I233> }; // XF86MonBrightnessUp
+    key <FK08> { overlay1=<MUTE> };
+    key <FK09> { overlay1=<VOL-> };
+    key <FK10> { overlay1=<VOL+> };
+
+    key <AE01> { overlay1=<FK01> };
+    key <AE02> { overlay1=<FK02> };
+    key <AE03> { overlay1=<FK03> };
+    key <AE04> { overlay1=<FK04> };
+    key <AE05> { overlay1=<FK05> };
+    key <AE06> { overlay1=<FK06> };
+    key <AE07> { overlay1=<FK07> };
+    key <AE08> { overlay1=<FK08> };
+    key <AE09> { overlay1=<FK09> };
+    key <AE10> { overlay1=<I236> }; // KEY_KBDILLUMTOGGLE
+    key <AE11> { overlay1=<I237> }; // KEY_KBDILLUMDOWN
+    key <AE12> { overlay1=<I238> }; // KEY_KBDILLUMUP
+    key <BKSP> { overlay1=<DELE> };
+
+    key <RTSH> { overlay1=<CAPS> };
+};
--- a/rules/base.m_s.part
+++ b/rules/base.m_s.part
@@ -8,7 +8,6 @@
   btc5126t      =       +inet(nav_acpi_common)
   btc9116u      =       +inet(btc9019u)
   btc6301urf    =       +inet(media_nav_acpi_common)
-  chromebook    =       +inet(chromebook)
   dellsk8125    =       +inet(dell)
   dellsk8135    =       +inet(dell)
   cherrybunlim  =       +inet(nav_common)
--- a/symbols/inet
+++ b/symbols/inet
@@ -1877,27 +1877,6 @@
 };
 
 partial alphanumeric_keys
-xkb_symbols "chromebook" {
-	include "level3(ralt_switch)"
-        key <FK01> {    [ XF86Back ] };
-        key <FK02> {    [ XF86Forward ] };
-        key <FK03> {    [ XF86Reload ] };
-        key <FK04> {    [ F11 ] };
-//        key <FK05> {    [ F5, F5, F5, F5 ] }; // Overview key
-        key <FK06> {    [ XF86MonBrightnessDown ] };
-        key <FK07> {    [ XF86MonBrightnessUp ] };
-        key <FK08> {    [ XF86AudioMute ] };
-        key <FK09> {    [ XF86AudioLowerVolume ] };
-        key <FK10> {    [ XF86AudioRaiseVolume ] };
-        key <BKSP> {    [ BackSpace, BackSpace, Delete ] };
-        key <UP>   {    [ Up, Up, Prior, Up ] };
-        key <DOWN> {    [ Down, Down, Next, Down ] };
-        key <LEFT> {    [ Left, Left, Home, Left ] };
-        key <RGHT> {    [ Right, Right, End, Right ] };
-        key <LWIN> {    [ Super_L, Super_L, Caps_Lock, Super_L ] };
-};
-
-partial alphanumeric_keys
 xkb_symbols "teck227" {
     include "inet(media_nav_common)"
     key <HIRA> { [ Hiragana ] };
--- a/symbols/Makefile.in
+++ b/symbols/Makefile.in
@@ -396,7 +396,8 @@
 ua us uz vn \
 za \
 altwin capslock compose ctrl empty eurosign rupeesign group inet \
-keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo
+keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo chromebook \
+chromebook_m chromebook_m_ralt chromebook_ralt
 
 EXTRA_DIST = $(symbols_DATA)
 all: all-recursive
