--- a/compat/Makefile.am
+++ b/compat/Makefile.am
@@ -1,7 +1,7 @@
 compatdir = $(xkb_base)/compat
 
 compat_DATA = \
-accessx basic caps complete \
+accessx basic caps complete chromebook \
 iso9995 \
 japan ledcaps \
 lednum ledscroll level5 \
--- a/symbols/Makefile.am
+++ b/symbols/Makefile.am
@@ -34,6 +34,6 @@
 ua us uz vn \
 za \
 altwin capslock compose ctrl empty eurosign rupeesign group inet \
-keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo
+keypad kpdl level3 level5 nbsp olpc shift srvr_ctrl typo chromebook
 
 EXTRA_DIST = $(symbols_DATA)
--- a/rules/base.ml1_c.part
+++ b/rules/base.ml1_c.part
@@ -1 +1,2 @@
+  chromebook	*		=	complete+chromebook
   *		*		=	complete
--- a/rules/base.xml.in
+++ b/rules/base.xml.in
@@ -4,6 +4,13 @@
   <modelList>
     <model>
       <configItem>
+        <name>chromebook</name>
+        <_description>Chromebook</_description>
+        <vendor>Google</vendor>
+      </configItem>
+    </model>
+    <model>
+      <configItem>
         <name>pc101</name>
         <_description>Generic 101-key PC</_description>
         <vendor>Generic</vendor>
--- a/rules/evdev.m_s.part
+++ b/rules/evdev.m_s.part
@@ -1,3 +1,4 @@
+  chromebook    =   +inet(evdev)+chromebook(overlay)
   $evdevkbds    =   +inet(evdev)+inet(%m)
   applealu_jis  =   +inet(evdev)+macintosh_vndr/jp(alujiskeys)
   *             =   +inet(evdev)
--- a/rules/evdev.xml.in
+++ b/rules/evdev.xml.in
@@ -1302,6 +1302,13 @@
        <vendor>htc</vendor>
      </configItem>
     </model>
+    <model>
+      <configItem>
+       <name>chromebook</name>
+       <description>Chromebook</description>
+       <vendor>Google</vendor>
+      </configItem>
+    </model>
   </modelList>
   <layoutList>
     <layout>
